### DISCUSSION TITLE
Supporting basePath applications with locales that exist before the basePath

### DISCUSSION THREAD

**Discussion by jaybytez** [vercel/next.js#22561] [Ideas]
**Describe the feature you'd like to request**

https://github.com/vercel/next.js/discussions/17078#discussioncomment-107903

As described in this comment list, we have multiple Next.js applications that co-exist with our other CMS Application to provide our overall website.  So Next.js runs a few apps/pages of the overall website.  Therefore in order to run Next.js applications, we have to use basePath, but the locale exists outside of the basePath.  I will use Robert's comment because it best represents our problem statement:

We have a site with multiple independent Next.js applications running at various base paths, e.g., /app-a and /app-b. It would be great for end users to see consistent example.com/[locale]/... paths for the whole site. Would this RFC then make these sites available at example.com/[locale]/app-a and example.com/[locale]/app-b

So we would like to use the i18n OOTB support with the locale being outside of the basePath.

**Describe the solution you'd like**

Possible a way to configure the locale location within the url that can be provided in next.config.js.

That way we can say within the examples above (example.com/[locale]/app-a and example.com/[locale]/app-b), that our locale can be extracted from the path previous to the basePath.  I assume this is the usual use case.

**Describe alternatives you've considered**

To use basePath with SSG support in order to get access to a locale prior to the basePath, we would have to use rewrites to handle that. So we would have to use rewrites to get access to the path so that we could rewrite example.com/[locale]/app-a as example.com/app-a/[locale].  This actually has a non-i18n bug where when parameters are in the url, the rewrite will create a link as /basePath/<route>/basePath. We also have to build a custom component to wrap links to ensure if we rewrite pages from /locale/basePath to /basePath/locale to use the i18n support, we have to make sure users still see the url as /locale/basePath. As well as the fact that we cannot go the direction of 'next export' with the rewrites.

The other direction is to dump SSG and use SSR to get access to the req so we can just take the locale and send it onto the Headless CMS that is doing all of the localization.  For us, we have complicated crosslinking that gets provided by our Headless CMS, so all we really need is the locale from the URL so we can send the locale to our Headless CMS API to receive the different versions of the content.  We can do this with SSR without any complex restructuring the project, but no ability to move to SSG.

The final direction, which we are taking, is to just remove basePath. In order to do this, we have to hard-code the basePath into any links, also use basePath in the assetPrefix, write quite a few rewrites for _next and api requests and then customize our Build ID to include the basePath and then write reverse proxy rules to use any URL with the basePath to route to the proper Next.js instance as we are running multiple sub-applications of the overall large site. Again, the downside is we cannot move the 'next export' direction of generating a static site because of all the rewrites we have to do with the basePath removed.

If there was a way to either get access to the locale prior to the basePath to utilize in API calls or better to support the Next.js OOTB i18n functionality with a locale that exists previous to the basePath.

**Comment by simpsonphile** [comment]
any update on this?

**Comment by vvxhid** [comment]
+1

**Comment by sainisagar310** [comment]
+1

**Comment by Muskos** [comment]
+1

**Comment by robertwbradford** [comment]
I would love to see something like the following syntax when `basePath` is used with i18n routing: https://github.com/vercel/next.js/discussions/17078#discussioncomment-220722
